cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1241,6 +1245,13 @@ class S3FileSystem::Impl {
     return std::string(FromAwsString(builder_.config().region));
   }
 
+  template <typename Error>
+  void SaveBackend(const Aws::Client::AWSError<Error>& error) {
+    if (!backend_ || *backend_ == S3Backend::Other) {
```

Comment:
Do we need this condition?

Reply:
Not necessarily, but if the backend was already detected, we probably don't want to undo the detection if the HTTP header sometimes misses or is incorrect.